### PR TITLE
[12.0][ADD] hr_attendance_sheet

### PR DESCRIPTION
**Pull request by patrickrwilson**
Adds HR Attendance Sheet module.

This module expands hr_attendance to incorporate sheets. An attendance sheet has a period of time specified in the general settings and incorporates the attendance entries that fall within that period. 

This module incorporates some sheet features:
- Sheets have an approval process also defined in the general settings. Approvals can be done by HR, Managers or HR and Managers.
- Auto Lunch is an option to calculate a lunch period if the attendance exceeds a specified duration.
- [UPDATE] Removed to put into own module] Split Overnight Attendance is an option to split the attendance at midnight if it crosses overnight. This is useful to prevent an attendance from crossing into another sheet period.
- [UPDATE] Added an 'Attendance Admin' field on departments to allow an employee to be assigned to a department that would administer that departments employee timesheets. This is in addition to the manager who can already review/approve them.

Credit is due to the author/maintainers of the timesheet_sheet module as much of the code in that module was referenced to create this one since it had similar functionalities: Miquel Raïch, Andrea Stirpe, Lois Rilo, Alexey Pelykh, Dennis Sluijk

### CODE REVIEW

**Comment by patrickrwilson** [comment]
Ping @max3903

**Comment by jarroyomorales** [comment]
@patrickrwilson Splitting attendances at 23:59 is a very interesting feature that helps a lot to manage night shifts and I think it could be extracted from this module and added onto an independent module.
Then we can make this module depend on the splitting module.

What do you think? I can work on it if you want

**Comment by patrickrwilson** [comment]
@jarroyomorales thanks for the feedback and yes, I think it's a good idea to put that feature into it's own module. I'll go ahead and remove that from this module, and feel free to generate the separate one. If you need help, let me know and ping me on the PR once you have it so I can review. Thanks again.

**Comment by murtuzasaleh** [comment]
@patrickrwilson Can you please remove `datetime.timedelta` and `datetime.time` if it is not used?

https://travis-ci.org/github/OCA/hr/jobs/710782871#L319
https://travis-ci.org/github/OCA/hr/jobs/710782871#L320

**Comment by patrickrwilson** [comment]
@murtuzasaleh I've removed those unused timedelta, time. Thanks.

**Comment by patrickrwilson** [comment]
@jarroyomorales did you want to take a stab at the split attendance module? If not I can go for it, I just don't want to duplicate efforts.

**Comment by patrickrwilson** [comment]
@jarroyomorales please see #875 for the separate split attendance module.

File: hr_attendance_sheet/models/hr_attendance.py [hunk 1]
```diff
@@ -0,0 +1,139 @@
...
+                    attendance.attendance_sheet_id = \
+                        attendance_sheet_ids or False
+
+    @api.multi
+    def _compute_duration(self):
+        for rec in self:
+            if rec.check_in and rec.check_out:
+                start = fields.Datetime.from_string(rec.check_in)
+                end = fields.Datetime.from_string(rec.check_out)
+                delta = end - start
```
Comment (Saran440) [inline hr_attendance_sheet/models/hr_attendance.py]:
```suggestion
                delta = rec.check_out - rec.check_in
```

rec.check_in and check_out is type datetime.
so no more fields.Datetime.from_string is needed

**Review by Saran440 — COMMENTED** [review]
Functional Test
Should it can't check-in if we lock state on attendance sheet?

My step : create attendance sheet and lock it.
Selection_007

Check-in
Selection_008

Check-out
Selection_009


and Code reviewed

**Comment by patrickrwilson** [comment]
@Saran440 Thank you, this was a good catch. I just added an IF so it will not link the attendance to a sheet if sheet has been locked. I figured this would be a better option than to prevent creating. I think this scenario would be a one off but it's good to account for it so thank you again.

**Comment by patrickrwilson** [comment]
@osi-scampbell @max3903 @Saran440 @murtuzasaleh This module is ready for further review. Thanks

File: hr_attendance_sheet/__manifest__.py [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+# License AGPL-3 - See http://www.gnu.org/licenses/agpl-3.0.html
+
+{
+    "name": "HR Attendance Sheet",
+    "version": "12.0.1.0.0",
+    "category": "Human Resources",
+    "summary": "Group attendances into attendance sheets.",
+    "website": "https://github.com/OCA/hr",
+    "author": "Odoo S.A.," "Odoo Community Association (OCA)",
+    "license": "LGPL-3",
```
Comment (Saran440) [inline hr_attendance_sheet/__manifest__.py]:
```suggestion
    "license": "AGPL-3",
```

This module is AGPL or LGPL?

File: hr_attendance_sheet/__manifest__.py [hunk 3]
```diff
@@ -0,0 +1,30 @@
+# Copyright 2020 Pavlov Media
+# License AGPL-3 - See http://www.gnu.org/licenses/agpl-3.0.html
+
+{
+    "name": "HR Attendance Sheet",
+    "version": "12.0.1.0.0",
+    "category": "Human Resources",
+    "summary": "Group attendances into attendance sheets.",
+    "website": "https://github.com/OCA/hr",
+    "author": "Odoo S.A.," "Odoo Community Association (OCA)",
```
Comment (Saran440) [inline hr_attendance_sheet/__manifest__.py]:
```suggestion
    "author": "Odoo S.A., Odoo Community Association (OCA)",
```

File: hr_attendance_sheet/models/hr_attendance.py [hunk 4]
```diff
@@ -0,0 +1,185 @@
+# Copyright 2020 Pavlov Media
+# License LGPL-3 - See http://www.gnu.org/licenses/lgpl-3.0.html
```
Comment (Saran440) [inline hr_attendance_sheet/models/hr_attendance.py]:
```suggestion
# License LGPL-3 - See http://www.gnu.org/licenses/lgpl-3.0.html
```

I'm not sure this module is AGPL or LGPL?

**Review by Saran440 — COMMENTED** [review]
Code Reviewed.

File: hr_attendance_sheet/__manifest__.py (hunk 2 above)
Comment (patrickrwilson) [reply hr_attendance_sheet/__manifest__.py]: It's AGPL, I fixed all the references to LGPL and change it. Thanks.

File: hr_attendance_sheet/__manifest__.py (hunk 3 above)
Comment (patrickrwilson) [reply hr_attendance_sheet/__manifest__.py]: Fixed

File: hr_attendance_sheet/models/hr_attendance.py (hunk 4 above)
Comment (patrickrwilson) [reply hr_attendance_sheet/models/hr_attendance.py]: It's AGPL, I fixed all the references to LGPL and change it. Thanks.

**Review by Saran440 — APPROVED** [review]
Code Reviewed.

**Comment by albariera** [comment]
When doing a functional test I saw that is possible to change fields as the employee_id, date_from, and date_to and remove lines when the sheet is approved or locked. Should it be like this?

Captura de pantalla de 2021-07-22 11-54-45

imagen

**Pull request closed without merging**